From fe577fb39f7481c58981ef788b27782bb543e863 Mon Sep 17 00:00:00 2001
From: Erik Schumacher <erik.schumacher@iris-sensing.com>
Date: Wed, 23 Oct 2024 09:36:12 +0000
Subject: [PATCH] pwm: imx-tpm: Use correct MODULO value for EPWM mode

The modulo register defines the period of the edge-aligned PWM mode
(which is the only mode implemented). The reference manual states:
"The EPWM period is determined by (MOD + 0001h) ..." So the value that
is written to the MOD register must therefore be one less than the
calculated period length.
A correct MODULO value is particularly relevant if the PWM has to output
a high frequency due to a low period value.

Upstream-Status: Backport [cc6a931d1f3b, probably in 6.13 and in patch
versions for earlier releases]

Signed-off-by: Erik Schumacher <erik.schumacher@iris-sensing.com>
---
 drivers/pwm/pwm-imx-tpm.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/pwm/pwm-imx-tpm.c b/drivers/pwm/pwm-imx-tpm.c
index a3a1342e29a4..da846dcef4b9 100644
--- a/drivers/pwm/pwm-imx-tpm.c
+++ b/drivers/pwm/pwm-imx-tpm.c
@@ -107,7 +107,9 @@ static int pwm_imx_tpm_round_state(struct pwm_chip *chip,
 	p->prescale = prescale;
 
 	period_count = (clock_unit + ((1 << prescale) >> 1)) >> prescale;
-	p->mod = period_count;
+	if (period_count == 0)
+		return -EINVAL;
+	p->mod = period_count - 1;
 
 	/* calculate real period HW can support */
 	tmp = (u64)period_count << prescale;
-- 
2.47.0

